### DISCUSSION TITLE
ForEachReducer can't get `idsBefore` for enum State

### DISCUSSION THREAD

**Discussion by Otbivnoe** [pointfreeco/swift-composable-architecture#3248] [Q&A] [answered]
As I understand the problem it tries to get `idsBefore`, but since the initial state is `empty (TripsChild.State.empty)` and key path refers to `TripsChild.State.trips`, it returns nil - that causes a crash after a force unwrap.

<img width="1073" alt="image" src="https://github.com/user-attachments/assets/368d9ff3-09b9-40bd-9ac2-7cdf5684148f">

<img width="1076" alt="image" src="https://github.com/user-attachments/assets/15293473-5318-460e-a8f6-ae82e6809ce6">

I wrote a simplified version of my code for a faster reproduce.

**Pay attention** it's important to pass `empty` as an initial state.

---

Starting point of the app:

```swift
TripsChildView(
    store: .init(initialState: .empty) {
        TripsChild()
    }
)
```

Content is the following: 

```swift
struct TripsChild: Reducer {

    @ObservableState
    @CasePathable
    @dynamicMemberLookup
    enum State: Equatable {
        case empty
        case trips(IdentifiedArrayOf<TripChildListItem.State>)
    }

    @CasePathable
    enum Action {
        case onAppear
        case trips(IdentifiedActionOf<TripChildListItem>)
    }

    var body: some Reducer<State, Action> {
        Reduce { state, action in
            return .none
        }
        .forEach(\.trips, action: \.trips) {
            TripChildListItem()
        }
    }
}

struct TripsChildView: View {

    let store: StoreOf<TripsChild>

    var body: some View {
        VStack {
            switch store.state {
            case .empty:
                EmptyView()
            case .trips:
                List {
                    ForEach(store.scope(state: \.trips, action: \.trips)) { store in
                        TripChildItemView(store: store)
                    }
                }
            }
        }
        .onAppear {
            store.send(.onAppear)
        }
    }
}
```

and a list item:

```swift
struct TripChildListItem: Reducer {

    @ObservableState
    struct State: Identifiable, Equatable {
        let id = UUID()
    }

    @CasePathable
    enum Action {}

    var body: some Reducer<State, Action> {
        EmptyReducer()
    }
}

struct TripChildItemView: View {

    let store: StoreOf<TripChildListItem>

    var body: some View {
        Text(store.state.id.uuidString)
    }
}

```

Xcode 15.4
Library version: 1.11.2

p.s. I asked the similar question yesterday (https://github.com/pointfreeco/swift-composable-architecture/discussions/3245), but seems like it's another problem.

**Comment by mbrandonw** [comment]
Hi @Otbivnoe, this is the same problem as #3245. The getters you are using are unavailable, and it was a bug in Swift that this code was allowed. It has been fixed in Swift 6.

**Reply by Otbivnoe** [reply, marked as answer]
Answer in this thread:  https://github.com/pointfreeco/swift-composable-architecture/discussions/3245
